pkg/controller/csi_handler.go

```diff
@@ -93,6 +101,99 @@ func (h *csiHandler) Init(vaQueue workqueue.RateLimitingInterface, pvQueue workq
 	h.pvQueue = pvQueue
 }
 
+// ReconcileVA lists volumes from the CSI Driver and reconciles the attachment
+// status with the corresponding VolumeAttachment object. If the attachment
+// status of the volume is different from the state on the VolumeAttachment the
+// VolumeAttachment object is patched to the correct state.
+func (h *csiHandler) ReconcileVA() error {
+	ctx, cancel := context.WithTimeout(context.Background(), h.timeout)
+	defer cancel()
+
+	// Loop over all volume attachment objects
+	vas, err := h.vaLister.List(labels.Everything())
+	if err != nil {
+		return errors.New("failed to list all VolumeAttachment objects")
+	}
+
+	// List volumes after getting volume attachment objects to prevent race
+	// condition where actual state is changed between ListVolumes and
+	// VolumeAttachment List. When listing VolumeAttachments first we can
+	// guarantee any race conditions will result in API object UPDATE failures
+	// instead of stomping on the VolumeAttachment with outdated data
+	published, err := h.volumeLister.ListVolumes(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to ListVolumes: %v", err)
+	}
+
+	for _, va := range vas {
+		nodeID, ok := va.Annotations[vaNodeIDAnnotation]
```

Comment:
When does nodeID move from annotation to spec?

Reply:
`VA.Spec.NodeName` is different from `NodeID` I think. NodeName refers to the Kubernetes node name - node ID is the reference for that node given by the CSI Driver.... I think :) In PD case the `nodeID` is fully qualified with project and zone, but the node name is not.

@msau42 in case I'm misrepresenting this